### DISCUSSION TITLE
Clarity on server actions action Ids changes in next.js 15

### DISCUSSION THREAD

**Discussion by shaunwarman** [vercel/next.js#73098] [App Router]
From the new [next.js 15 blog release](https://nextjs.org/blog/next-15) it looks like in the [server actions security section](https://nextjs.org/blog/next-15#enhanced-security-for-server-actions) there is a snippet saying

>Secure action IDs: Next.js now creates unguessable, non-deterministic IDs to allow the client to reference and call the Server Action. These IDs are periodically recalculated between builds for enhanced security.

Does this mean that the IDs will be recalculated at runtime and change in production? Or per build? And can there be some added detail on how this is adding security (maybe security through obscurity?). I believe the action IDs are quite random today (just a checksum of the code AFAIK).

**Comment by knpwrs** [comment]
[I asked about this on reddit](https://www.reddit.com/r/nextjs/comments/1i7hx08/next_14_an_honest_review/ma0xe2k/?context=3) and was pointed here: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#overwriting-encryption-keys-advanced

It would be great to see some documentation on how and when the endpoint IDs change. It would be a shame to deploy a change and break clients that haven't refreshed yet.

**Reply by shaunwarman** [reply]
Right,  a bit concerning. Seems like another friction point for self hosting / non vercel deployments.

**Comment by zoolyka** [comment]
When `NEXT_SERVER_ACTIONS_ENCRYPTION_KEY` is specified, the `next-action` ID remains consistent across server instances and builds, even if the action's source code is modified. It likely just hashes the function name with the provided encryption key _(which can be generated with `openssl rand -base64 32`)_. I believe that if your actions have reasonable security measures in place, it is unnecessary to let Next.js rotate these IDs automatically.

**Reply by stanplaat** [reply]
> When `NEXT_SERVER_ACTIONS_ENCRYPTION_KEY` is specified, the `next-action` ID remains consistent across server instances and builds, even if the action's source code is modified. It likely just hashes the function name with the provided encryption key _(which can be generated with `openssl rand -base64 32`)_. I believe that if your actions have reasonable security measures in place, it is unnecessary to let Next.js rotate these IDs automatically.

have you success with this scenario ? unfortunatelly not for me. I am using multiple server for load balancer, and this action id's broke up in my apps.

**Reply by zoolyka** [reply]
Yes, I'm using Next 15 and placed this variable inside the /.env file.

**Reply by tomdohnal** [reply]
@stanplaat make sure the `NEXT_SERVER_ACTIONS_ENCRYPTION_KEY` env var is present during build time (that's when the action ids are generated). i had made the mistake of only adding in runtime and that doesn't work

**Reply by markdemich-fl** [reply]
I set the variable at build time and it seems to work per environment.  However, since it's at build time I expected the IDs to be the same across my environments (i.e., QA, Stage, Prod).  However, when I compare the same operation in QA and Stage they seem to have different ids.  Is this expected?  I was hoping they would be the same since I'm writing a few load test scripts that need the ID to be stable across environments.  Should I also be setting the environment variable at runtime?

**Reply by austinb847** [reply]
anyone know if you deploy on vercel, do they automatically handle this? or do you still need to use the key to get this working as expected? my app recently ran into this issue. deployed to vercel and had a ton of runtime logs eg:

`[Error: Failed to find Server Action. This request might be from an older or newer deployment. Original error: Cannot read properties of undefined (reading 'workers')]`
